### PR TITLE
feat: Add installation deviceToken deduplication options

### PR DESCRIPTION
**Pull request by mtrezza**
## Pull Request

- Report security issues [confidentially](https://github.com/parse-community/parse-server/security/policy).
- Any contribution is under this [license](https://github.com/parse-community/parse-server/blob/alpha/LICENSE).

## Issue

Parse Server's `_Installation` deduplication on `deviceToken` collisions today runs as master and always destroys the conflicting row, regardless of the caller's auth context or the developer's intent. This means:

1. ACL and CLP set on `_Installation` are not honored on the dedup path — the dedup bypasses them silently.
2. The conflicting row is destroyed wholesale, losing any custom fields, channels, or history stored on it.
3. In the merge case (where the new install matches one row by `installationId` and another by `deviceToken` only), the deviceToken-only orphan always wins, even when the active install carries more meaningful data.

Developers have no way to opt out of any of these behaviors.

## Approach

Add a new `installation` server option block with three options:

- `installation.duplicateDeviceTokenActionEnforceAuth: Boolean` — default `false`. When `true`, the dedup operation runs with the caller's auth context so ACL and CLP are honored. Master and maintenance keys always bypass regardless. The default `false` is registered as deprecated (DEPPS24); the planned default in 10.0.0 is `true`.
- `installation.duplicateDeviceTokenAction: 'delete' | 'update'` — default `'delete'`. `'update'` clears the now-conflicting ID field on the conflicting row instead of destroying it, preserving custom fields, channels, and history.
- `installation.duplicateDeviceTokenMergePriority: 'deviceToken' | 'installationId'` — default `'deviceToken'`. Selects which side wins the merge case (when the new install matches an existing row by `installationId` and another by `deviceToken` without an `installationId`).

Implementation:

- New module `src/InstallationDedup.js` with two helpers (`removeConflictingDeviceToken` for new-install/PUT-deviceToken-change collisions, `applyDuplicateDeviceTokenMerge` for the merge case). Both honor the option-driven dispatch and log per a uniform policy (`verbose` for success / no-match, `warn` for permission denial, `error` for unexpected errors).
- `RestWrite.handleInstallation` refactored at three sites to call the helpers. All three sites are now consistently awaited; today's two new-install / PUT sites were fire-and-forget. The latency cost is bounded to requests that actually fire dedup.
- `Config.validateInstallations` validates the new option block; throws on unknown keys, wrong types, or invalid enum values; defaults missing nested keys.
- `DEPRECATIONS.md` entry DEPPS24 for the future default flip on `duplicateDeviceTokenActionEnforceAuth`.
- README section under Configuration documenting all three options with examples.

Default option values preserve today's behavior, so this is non-breaking for existing apps.

## Tasks

- [x] Add tests
- [x] Add changes to documentation (guides, repository pages, code comments)

## Summary by CodeRabbit

* **New Features**
  - Added configurable `installation` options to control duplicate device-token handling: action (delete/update), auth enforcement, and merge priority.

* **Refactor**
  - Centralized installation dedup/merge logic into shared helpers and updated write paths to use them.

* **Configuration**
  - Validation and defaults for the new `installation` settings; environment-driven option surface exposed.

* **Documentation / Deprecations**
  - New "Installations" docs and a deprecation announcing the upcoming default change for auth enforcement.

* **Tests**
  - Comprehensive tests covering dedup, merge, auth enforcement, validation, and edge cases.

### CODE REVIEW

File: src/InstallationDedup.js [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+    if (err && err.code === Parse.Error.OBJECT_NOT_FOUND) {
+      logResult(action, 0, err);
+    } else if (err && err.code === Parse.Error.OPERATION_FORBIDDEN) {
+      logResult(action, null, err);
+    } else {
+      logResult(action, null, err);
+      throw err;
+    }
+  }
+  return survivorId;
```
[the remark this answers is not in the document]
Comment (mtrezza) [reply src/InstallationDedup.js:170]: Good catch. Confirmed reachable on PUT to `/installations/:objectId` when the existing row has the same `deviceToken` and no `installationId` — the merge picks `idMatch === deviceTokenMatch` and the destroy hits the survivor. Pre-existing in master too, but worth fixing here since we own this code path now. Applied the early-return guard in 7ce2ff39b with a unit test that fails against the unfixed code (`Expected spy destroy not to have been called`) and passes after the fix.

File: src/Options/docs.js [hunk 2]
```diff
@@ -148,6 +149,13 @@
...
+ * @property {Boolean} duplicateDeviceTokenActionEnforceAuth Whether the `_Installation` deduplication operation enforces the caller's auth context (and the resulting ACL and CLP). When `true`, the dedup `destroy`/`update` runs with the caller's `runOptions`, so ACL and CLP are honored. When `false`, the dedup runs as master and bypasses both. Master and maintenance keys always bypass regardless of this flag. Default is `false`.
+ * @property {String} duplicateDeviceTokenMergePriority At the merge case (when an existing row holds the new `deviceToken` but has no `installationId` of its own), which side wins. `'deviceToken'` — the deviceToken-only row survives, the request's `idMatch` row is the loser. `'installationId'` — the request's `idMatch` (active install) survives, the deviceToken-only orphan is the loser. Default is `'deviceToken'`.
+ */
```
[the remark this answers is not in the document]
Comment (mtrezza) [reply src/Options/docs.js:157]: Deprecation status is intentionally not mentioned in option help text per project convention; it lives exclusively in DEPRECATIONS.md (DEPPS24) and src/Deprecator/Deprecations.js. The option key itself is not slated for removal — only its default value flips in the next major. The "Deprecated." prefixes on a few existing options (e.g. playgroundPath, allowExpiredAuthDataToken) are inconsistent holdovers, not the canonical pattern.

File: src/InstallationDedup.js [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+  }
+  if (action === 'update') {
+    return database.update(
+      CLASS_NAME,
+      query,
+      { [fieldToClear]: { __op: 'Delete' } },
+      runOptions,
+      many,
+      validSchemaController
+    );
```
[the remark this answers is not in the document]
Comment (mtrezza) [reply src/InstallationDedup.js:47]: Confirmed. `DatabaseController.update` is `(className, query, update, { acl, many, upsert, addsField } = {}, skipSanitization = false, validateOnly = false, validSchemaController)`. The original call passed `many` as the 5th positional (`skipSanitization` slot) and `validSchemaController` as the 6th (`validateOnly` slot), which left the actual schema-controller slot empty and caused multi-row dedup-update operations to fall through to `findOneAndUpdate` (only one row updated). Fixed in 0c05d6bc8 by moving `many` into the options object and putting `validSchemaController` in the correct slot. Added an integration test that creates three rows sharing the same `deviceToken` via direct adapter insertion and asserts all three get cleared by the dedup — fails on the unfixed code, passes after the fix.

File: spec/ParseInstallation.spec.js [hunk 4]
```diff
@@ -1297,4 +1297,481 @@ describe('Installations', () => {
...
+      ).toBeRejectedWith('installation must be an object.');
+    });
+
+    it('should reject unknown nested keys', async () => {
+      await expectAsync(
+        reconfigureServer({
+          installation: { unknownKey: 'foo' },
+        })
+      ).toBeRejectedWith("installation contains unknown property 'unknownKey'.");
+    });
```
[the remark this answers is not in the document]
Comment (mtrezza) [reply spec/ParseInstallation.spec.js:1731]: This is incorrect. `src/Config.js:731` throws `` `installation contains unknown property '${key}'.` `` — exactly what the test expects. There is no occurrence of the string "is not a valid installation option" anywhere in the source. The test passes against the actual code (`1 spec, 0 failures`).

**Comment by mtrezza** [comment]
@coderabbitai resume

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [9.9.0-alpha.3](https://github.com/parse-community/parse-server/releases/tag/9.9.0-alpha.3)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [9.9.0](https://github.com/parse-community/parse-server/releases/tag/9.9.0)
